### DISCUSSION TITLE
Is fine to configure IdentityOption to use Subject as UserId?

### DISCUSSION THREAD

**Discussion by kamnab** [JonPSmith/AuthPermissions.AspNetCore#47] [Q&A]
I have integrated OpenIddict with AuthP and this led to use Subject as claim based for the authentication. However, some parts got errors that require to modify the claim type of NameIdentifier to Subject. I am not sure if this needs to be checked in somewhere else, pls advise

**Comment by JonPSmith** [comment]
Hi @kamnab,

I'm pleased to hear that you are integrated OpenIddict with AuthP, as that could be useful for others. But I will need some more information to help you with the errors you have. Please show your code and include the exception's stracktrace.

In the meantime here are some comments / links that might help you in the design of your OpenIddict server.

To use the AuthP library you need to:
1. Link the ASP.NET Core authenticated user to the AuthP user. This is done by taking the `ClaimTypes.NameIdentifier` claim (which I refer to as the _userId_) of the  ASP.NET Core authenticated user and using this userId when creating the AuthP.
2. You need a way to add the extra AuthP claims to the user's claims. Typically we intercept the login and add the AuthP claims (see the in the `AuthPRegion` in the [`AzureAdOpenIdExtension`](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/AuthPermissions.AspNetCore/OpenIdCode/AzureAdOpenIdExtension.cs) code). 

## Looking at OpenIddict

The problem is to link a new OpenIddict user to a AuthP user, with the AuthP Roles, Tenant etc.

I can see three possible approaches, but I don't understand OpenIddict well enough to say which will work.

1. The first is complex, but I'm sure it would work. In this case you would add the AuthP library to your OpenIddict server so that you can create an AuthP whenever a new user is added. As I said its complex so contact me if you want to go this way.
2. If you can add a feature in your OpenIddict server to allow your main application **find** a OpenIddict user by email and **create** a new OpenIddict user, then you can use the way the new [`AzureAdNewUserManager`](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/AuthPermissions.SupportCode/AddUsersServices/Authentication/AzureAdNewUserManager.cs) works - see the article [Three ways to securely add new users to an application using the AuthP library](https://www.thereformedprogrammer.net/three-ways-to-securely-add-new-users-to-an-application-using-the-authp-library/) for more.
3. If you can add a feature in your OpenIddict server to allow main application to **return all the OpenIddict users** you could use the "sync users" approach. Again see the article [Three ways to securely add new users to an application using the AuthP library](https://www.thereformedprogrammer.net/three-ways-to-securely-add-new-users-to-an-application-using-the-authp-library/) for more.

**Comment by kamnab** [comment]
Thank you for your suggestion. 

Let me provide you configuration I have made and error I got:

![image](https://user-images.githubusercontent.com/7078364/175250627-f03f7c61-1e6d-4e08-8ee4-994e48568f36.png)

![image](https://user-images.githubusercontent.com/7078364/175251850-83dfb05b-9973-4495-9f64-fcde51fc5b15.png)

This got error if I do not check claim type with x.Type == "sub"

**Reply by JonPSmith** [reply]
I now see what your problem is - the claims names don't match the normal ASP.NET Core claim names. I had this problem when working with Azure AD and I had to alter claims in the event handler - see [this section](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/AuthPermissions.AspNetCore/OpenIdCode/AzureAdOpenIdExtension.cs#L34L44) of the of the OpenIdConnect login event.

So in your situation you just have to  create a claim with the type of `ClaimTypes.NameIdentifier` for AuthP to work, and you can do that within the login event.
